src/main/java/org/quiltmc/loader/api/config/MetadataType.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.loader.api.config;
+
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * A typed key to be used for setting and getting object metadata.
+ */
+@ApiStatus.NonExtendable
+public interface MetadataType<T> {
+	String getId();
```

Comment:
I don't think that would happen often. I think raw string metadata should have a lightweight wrapper class anyway (for constants or helper methods).

Reply:
It would make sense to me to have a class passed *in addition to* the string ID, but it doesn't make sense to me to use that as identifying information. Having wrapper classes over String metadata also doesn't seem very useful to me.

Reply:
I think I've had the exact same argument with the attribute lookup api. I'm just a fan of using strings and magic constants when they're not needed.
I guess it mostly comes down to what metadata is used for. Other than comments.